src/api/source/mvt.js

```diff
@@ -0,0 +1,452 @@
+import Base from './base';
+import CartoValidationError from '../error-handling/carto-validation-error';
+import * as rsys from '../../client/rsys';
+import Dataframe from '../../core/dataframe';
+import * as util from '../util';
+import * as Protobuf from 'pbf';
+import * as LRU from 'lru-cache';
+import { VectorTile, VectorTileFeature } from '@mapbox/vector-tile';
+
+export default class MVT extends Base {
+
+    /**
+     * Create a carto.source.MVT.
+     *
+     * @param {object} data - A MVT data object
+     *
+     * @example
+     * new carto.source.MVT("https://{server}/{z}/{x}/{y}.mvt");
+     *
+     * @fires CartoError
+     *
+     * @constructor MVT
+     * @extends carto.source.Base
+     * @memberof carto.source
+     * @api
+     */
+    constructor(serverURL, metadata) {
+        super();
+        // this._validateServerURL(serverURL);
+        this._serverURL = serverURL;
+        this._oldDataframes = [];
+        const lruOptions = {
+            max: 1000
+            // TODO improve cache length heuristic
+            , length: function () { return 1; }
+            , dispose: (key, promise) => {
+                promise.then(dataframe => {
+                    if (!dataframe.empty) {
+                        dataframe.free();
+                        this._removeDataframe(dataframe);
+                    }
+                });
+            }
+            , maxAge: 1000 * 60 * 60
+        };
+        this.metadata = metadata;
+        this.cache = LRU(lruOptions);
+    }
+
+    _clone(){
+        return new MVT(this._serverURL, this.metadata);
+    }
+
+    bindLayer(addDataframe, removeDataframe, dataLoadedCallback) {
+        this._addDataframe = addDataframe;
+        this._removeDataframe = removeDataframe;
+        this._dataLoadedCallback = dataLoadedCallback;
+    }
+
+    async requestMetadata(viz) {
+        this._MNS = viz.getMinimumNeededSchema();
+        this._resolution = viz.getResolution;
+        return this.metadata;
+    }
+
+    requestData(viewport) {
+        const tiles = rsys.rTiles(viewport);
+        this._getTiles(tiles);
+    }
+
+    free() {
+    }
+
+    _getTiles(tiles) {
+        var completedTiles = [];
+        var needToComplete = tiles.length;
+        tiles.forEach(t => {
+            const { x, y, z } = t;
+            this._getDataframe(x, y, z).then(dataframe => {
+                if (dataframe.empty) {
+                    needToComplete--;
+                } else {
+                    completedTiles.push(dataframe);
+                }
+                if (completedTiles.length == needToComplete) {
+                    this._oldDataframes.map(d => d.active = false);
```

Comment:
There is a `requestGroupID` logic in `windshaft.js` that has not been placed here. That logic is important since without it, if tiles from an early request come after the last tiles requested are processed, then the first ones replace the new ones, which may not cover the screen completely.

Reply:
We probably should refactor and extract a `TiledSource` class to manage all this logic, then derive both `Windshaft` and `MVT` from it (or use it as a helper). But for now I'd just reproduce this exactly as in Windshaft and leave the refactor for after this is merged.

Reply:
added